speed up listDatasetFileDetails API

vkuznet opened the pull request:
Fixes #11098 

#### Status
ready

#### Description
Use concurrent features to speed up `listDatasetFileDetails` API. In my local setup I achieve speed up by factor of 5 or more for a specific dataset. Here are benchmark numbers for using `/VBF1Parked/Run2012D-v1/RAW` dataset. This dataset contains 594 blocks. The current codebase takes approximately 190 seconds to fetch its parents and file lumis. Using proposed solution I achieve the following numbers:
- 37 seconds using 50 concurrent tasks
- 41 seconds using 100 concurrent tasks
- 90 seconds using 10 concurrent tasks

Internally, I used `requests` python library instead of `pycurl_manager.py`. The latter is not suitable for concurrent execution since it is not thread safe as it holds global curl object. As such curl options are set on first tasks, but can't be changed (since code set them up) in others until first task is completed. The `requests` library does not depend on global curl object and curl library itself, and it is suitable for concurrent execution of multiple URL calls.

This PR only addresses speed up of single  `listDatasetFileDetails` DBS3Reader API but other APIs where multiple (sequential) calls to DBS are made can benefit from this approach.

#### Is it backward compatible (if not, which system it affects?)
YES

#### Related PRs
https://github.com/dmwm/dbs2go/issues/5

#### External dependencies / deployment changes
relies on Python `requests` library

vkuznet committed:
speed up listDatasetFileDetails API

vkuznet commented:
@amaltaro , @todor-ivanov , @klannon this PR speed up single (most expensive) DBS3Reader API by factor of 5 or more. The code shows that concurrent execution of calls to upstream server can provide significant improvement. I suggest that you study this pattern which can be applied to different parts of WMCore codebase where multiple calls to upstream server (DBS, Rucio, etc.) will be required. Even thought it is not in Q2 plan I decided to put this forward as I obtained significant improvements can be made which entire system can benefit from and (more importantly) will allow to work with large datasets and avoid timeouts on CMSWEB.

vkuznet commented:
@amaltaro , on my second commit I got weird unit test failure which I'm 100% sure is not related to my changes but I want to point you to it since it may show how this unit test became unstable. Please see https://cmssdt.cern.ch/dmwm-jenkins/job/DMWM-WMCore-PR-test/13026/testReport/junit/WMCore_t.Services_t.UUID_t/UUIDTest/testUUID/ which shows:
```
'bcb3' == 'bcb3' : Second component of UUID the same bcb3 != bcb3
```
I think it is random error since UUID may be close enough during random process generation. Moreover, the message of test is wrong. I suggest that we address this unit issue separately. I'll trigger test again.

vkuznet commented:
test this please

amaltaro reviewed (commented):
Valentin, even though this looks like a great performance boost, I do not think this is the right implementation.

Why don't we use the pycurl_manager module for this? We already have some DBS concurrent APIs implemented in this module (inherited from your work porting Unified features into WMCore):
https://github.com/dmwm/WMCore/blob/master/src/python/WMCore/MicroService/Tools/Common.py

and I think the best way forward would be to actually create a new module under:
WMCore/Services/DBS/youNameIt.py

and put this code in there, if possible relying on pycurl_manager only instead of `requests`.
If you see a need to use the straight forward `requests` library, then I think we could have a chat first before putting all this effort in.

Please let me know your thoughts (BTW, it's a holiday here, so you might hear back from me only on Monday).

vkuznet commented:
@amaltaro I explained in description that `pycurl_manager.py` can't be used in concurrent futures since it locks curl object. The only way to use `pycurl_manager` is to use `multi_getdata` and it is on my todo list.

vkuznet commented:
ok, I refactored the code using `pycurl_manager.py` `multi_getdata` and results are even more impressive. I achieve 10+ improvements for this API. Here is new structure:
- WMCore/Services/DBS/DBSUtils.py contains new common code
- WMCore/Services/DBS/DBS3Reader.py now can use either approach, tested with `requests` and `pycurl_manager.py` libraries
- I borrowed code from WMCore/MicroService/Tools/Common.py which seems to be common between these two code-bases, therefore I suggest to re-factor the code again to put this common code elsewhere. Please see appropriate comment in WMCore/Services/DBS/DBSUtils.py

Bottom line, I would like to keep around concurrent.features code since it demonstrates new paradigm on how concurrent programming should be done. This concept can be applied to different parts of WMCore which may benefit from it. The `requests` library used in original codebase is just an example and I suggest to keep it around, but it is not required now after my second implementation with `multi_getdata` from `pycurl_manager.py`

vkuznet commented:
New best value I achieved with `multi_getdata` is 16 seconds, which is almost factor of 12 better than current implementation based on sequential programming pattern.

vkuznet commented:
I re-factor the code and generalize it. The new API `dbsParallelApi` will allow to be used by different APIs in DBS3Reader (see full list in https://github.com/dmwm/WMCore/issues/11098). I added new attribute `parallel` into constructor of DBS3Reader such that class can execute either linear (sequential) or concurrent (parallel) DBS APis workflows. The code now can be further expanded to speed-up other APIs.

vkuznet commented:
I rebased this PR since we merged #11101

amaltaro commented on src/python/WMCore/Services/DBS/DBSUtils.py:
```diff
@@ -0,0 +1,149 @@
+    :param logger: logger
+    :return: list of file parents
+    """
+    urls = ['%s/fileparents?block_name=%s' % (dbsUrl, b.replace('#', '%23')) for b in blocks]
```
Instead of replacing string, I'd suggest to stick to the standard way of encoding urls (based on urllib). In this case, you want to use `urllib.parse.quote_plus`.

vkuznet replied:
that's fine

amaltaro commented on src/python/WMCore/Services/DBS/DBSUtils.py:
```diff
@@ -0,0 +1,149 @@
+    :param logger: logger
+    :return: list of file lumis
+    """
+    urls = ['%s/filelumis?block_name=%s' % (dbsUrl, b.replace('#', '%23')) for b in blocks]
```
Same comment as the one above.

amaltaro commented on src/python/WMCore/Services/DBS/DBSUtils.py:
```diff
@@ -0,0 +1,149 @@
+    :param logger: logger
+    :return: list of block origins for a given parent lfns
+    """
+    urls = ['%s/blockorigin?block_name=%s' % (dbsUrl, b.replace('#', '%23')) for b in blocks]
```
Same as the comment above.

amaltaro commented on src/python/WMCore/Services/DBS/DBSUtils.py:
```diff
@@ -0,0 +1,149 @@
+        lumis = urllib.parse.quote_plus(str(fileInfo['lumi_section_num']))
+        url = '%s/files?dataset=%s&run_num=%s&lumi_list=%s' \
+                % (dbsUrl, parentDataset, run, lumis)
+        if ':8443' not in url:
```
I might be wrong, but I think this switch has already been covered by the portForward decorator, no?

vkuznet replied:
this code does not use any decorators, and no, I checked, the port was not changed, that's why I added the change.

amaltaro replied:
Then I'd suggest to use the exact same mechanism that Todor implemented in the last year, as you can see here:
https://github.com/dmwm/WMCore/pull/10330/files

in short:
```
portForwarder = PortForward(8443)
url = portForwarder(url)
```

amaltaro commented on src/python/WMCore/Services/DBS/DBSUtils.py:
```diff
@@ -0,0 +1,149 @@
+        else:
+            key = url
+        code = int(row.get('code', 200))
+        if code != 200:
```
Shouldn't we consider fail http requests only if the status code is >= 400?

vkuznet replied:
this is an edge use-case since non 200 codes does not return any data and may return plain string, e.g. the code 301 will return "Moved permanently", and code below explicitly extracts the `data` which is constructed by `multi_getdata` function from `pycurl_manager`. I would leave it as is because it explicitly will show non-data message.

amaltaro replied:
If my memory doesn't fail me, I've seen cases where we get successfully get actual data and the return code is != 200. Perhaps we should then consider at least codes 200-200 as successful? From:
https://developer.mozilla.org/en-US/docs/Web/HTTP/Status

amaltaro reviewed (changes requested):
Valentin, this is looking pretty good in general.

However, I left some comments along the code and I would like to see unit tests created for some of those functions in the DBSUtils module. Those actually making calls to the DBS server might have to be skipped though (unless we mock things).

Valentin, once you update it, I will have a final and thorough review.

vkuznet commented:
@amaltaro I added one _**integration**_ test to test functionality of `listDatasetFileDetails` DBS API through sequential and parallel execution. I found small dataset (in terms of blocks/files) in production DBS DB and used for the test. Please review and let me know if it is appropriate.

amaltaro commented on src/python/WMCore/Services/DBS/DBSUtils.py:
```diff
@@ -0,0 +1,149 @@
+import json
+import urllib
+from urllib.parse import urlparse
+from urllib.parse import parse_qs, quote_plus
```
These two imports should be grouped.

vkuznet replied:
fine

amaltaro commented on test/python/WMCore_t/Services_t/DBS_t/DBSUtils_t.py:
```diff
@@ -0,0 +1,67 @@
+        url = 'http://a.b.com?d=1&f=bla'
+        results = urlParams(url)
+        self.assertTrue(results.get('d'), 1)
+        self.assertTrue(results.get('f'), 'bla')
```
Maybe testing `results` with:
```
self.assertCountEqual(results, {'d': ['1'], 'f': ['bla']})
```
would be cleaner and better?

Can you please also add a test case for an url with multiple values for the same query string (e.g. d=1&d=2&f=bla)?

vkuznet replied:
done

amaltaro commented on test/python/WMCore_t/Services_t/DBS_t/DBSUtils_t.py:
```diff
@@ -0,0 +1,67 @@
+        return
+
+
+    def testGetParallel(self):
```
It's great to have a test like this, even if it's not run in jenkins. At least we can make sure our code is sound!

To avoid triggering this by every single jenkins run, can you please:
```
from nose.plugins.attrib import attr
```

and decorate this test with:
```
@attr("integration")
```
?

vkuznet replied:
ok

amaltaro reviewed (changes requested):
Thank you for these changes, Valentin. I have some extra comments for your consideration.

vkuznet commented:
@amaltaro , I adjusted all comments you made except codes. The codes 200-300 have special meaning, e.g. 201 Created  is used to create resource, and we are not creating anything, 202 Accepted The request has been received but not yet acted upon, so when we query the data we should receive the data, etc. Please do not mix codes with actual behavior. Each code mans something and our action ONLY to receive the data which falls into code 200. All other codes are not appropriate for our action. That's why code should not use them as correct condition.

vkuznet commented:
@amaltaro , @todor-ivanov what is holding us here? I addressed all issues you asked for, and already other PR changed DBS3Reader that I was need to resolve conflict.

amaltaro reviewed (changes requested):
Valentin, you can find a couple of comments along the code.

In addition to those, I have some other remarks to make:
* maybe we should rename the DBSUtil module to something more meaningful, e.g.: DBSParallel, or DBSPycurl, or DBSConc? Just to give a clear distinction that its underlying library is different than DBS3Reader
* logger parameter in those DBSUtil functions is not consistent. Sometimes it's mandatory, others it's optional. From a different angle, maybe it could be resolved by converting those functions into methods of a class, defining the logger object only once when it gets instantiated. Just an idea though...
* the DBSUtil functions are not consistent as well in the way they got developed. For instance, `dbsListFileParents` and other 2 only return data, while `dbsParentFilesGivenParentDataset` replicates the whole logic from DBS3Reader. I would stick to one model only, so either we (re-)implement everything in the DBSUtil function, or we only retrieve data and send it not processed back to DBS3Reader
* these changes are not compliant with the guidelines. New modules should not have any pylint/pep8/pycodestyle issues, unless there is no way around of it (as mentioned [here](https://github.com/dmwm/WMCore/blob/master/CONTRIBUTING.rst#automatic-tests))
* there is a mix of variable names as well (included in this PR, what was there already does not need to be touched), e.g. `block_parents`.

It's good that Erik is working on the pep8/pycodestyle, so we can at least get these reports now and fix it before human review have to spend time on it.

amaltaro commented on test/python/WMCore_t/Services_t/DBS_t/DBSUtils_t.py:
```diff
@@ -0,0 +1,75 @@
+
+        """
+        url = "https://cmsweb.cern.ch/dbs/prod/global/DBSReader"
+        reader1 = DBS3Reader(url, logger=None, parallel=False, aggregate=True)
```
Valentin, I like how you made this integration test. Can you please do the same for the other APIs that you offer a concurrent option (please keep them marked as integration, it's really only to make sure all these changes are providing the same result and to help testing whenever we need to change it again).

vkuznet commented on src/python/WMCore/Services/DBS/DBS3Reader.py:
```diff
@@ -307,6 +310,25 @@ def listDatasetFileDetails(self, datasetPath, getParents=False, getLumis=True, v
             files[f['logical_file_name']]['Parents'] = []
 
+        # parallel execution for listFileParents and listFileLumis APIs
+        if self.parallel:
```
Your changes do two things:
- they modify existing code base, since you add new loop to create block_parents dict
- they create new dict (which adds to the memory consumption)

In my version, I did not change existing codebase.

It is a choice and I don't have strong preference in one way or another. I leave it up to you to decide. Bottom line, the concurrent/parallel construct in a code quite often require new redesign and it is hard to align them with sequential programming patterns.

amaltaro replied:
The memory concern is a good point. However, if we use the concurrent version, then we will end up with that large dictionary anyways.

As anything, there are pros and cons and it's hard to decide whether a) making a transparent change now is better than b) making the code maintenance easier for future changes. Given that you have already coded and tested it, I'd say we keep the current implementation.

vkuznet commented:
Alan, let me answer your questions in order they appear. For clarify I'll repeat the list:
- maybe we should rename the DBSUtil module to something more meaningful, e.g.: DBSParallel, or DBSPycurl, or DBSConc? Just to give a clear distinction that its underlying library is different than DBS3Reader
   - VK: Please be specific, I choose name using my view, if you want to name it differently please specify how, I should not guess which name out of your suggestion is more appropriate
- logger parameter in those DBSUtil functions is not consistent. Sometimes it's mandatory, others it's optional. From a different angle, maybe it could be resolved by converting those functions into methods of a class, defining the logger object only once when it gets instantiated. Just an idea though...
   - VK: indeed, logger parameter seems inconsistent, I only added based on existing DBS3Reader code. But after your review I decided to completely remove it as it is not used by parallel code at all
- the DBSUtil functions are not consistent as well in the way they got developed. For instance, dbsListFileParents and other 2 only return data, while dbsParentFilesGivenParentDataset replicates the whole logic from DBS3Reader. I would stick to one model only, so either we (re-)implement everything in the DBSUtil function, or we only retrieve data and send it not processed back to DBS3Reader
   - VK: this is side-effect of code review, originally I placed code within DBS3Reader which then you requested to put into new module. As such code was copied to new module and inherited original naming convention used in DBS3Reader
- these changes are not compliant with the guidelines. New modules should not have any pylint/pep8/pycodestyle issues, unless there is no way around of it (as mentioned [here](https://github.com/dmwm/WMCore/blob/master/CONTRIBUTING.rst#automatic-tests))
   - VK: fair enough, will improve
- there is a mix of variable names as well (included in this PR, what was there already does not need to be touched), e.g. block_parents.
   - VK: I made changes consistent with names used by DBS3Reader. I don't know what is better to have consistency with code surrounding my changes or be strict with guidelines to make new code. In latter case, new code will be inconsistent with existing code. I am open to suggestion here. If you insist to make new code based on guidelines I'll adjust new variables but at the end we'll end-up with a mix of naming conventions. Even your proposal to adjust code violates naming conventions. Bottom line, guidelines are not a panacea to all cases.

vkuznet commented:
test this please

vkuznet commented:
@amaltaro , I made necessary changes and replied to all your comments (some of them were unclear and I left my questions). Meanwhile, I run code through `autopep8`, and verified that Jenkins reports 10 score for new module. I also complement the code with integration tests. Please review again.

amaltaro commented on src/python/WMCore/Services/DBS/DBS3Reader.py:
```diff
@@ -72,17 +74,18 @@ class DBS3Reader(object):
     """
 
-    def __init__(self, url, logger=None, **contact):
+    def __init__(self, url, logger=None, parallel=None, **contact):
```
Can you please create a docstring for this class and specify what `parallel` is for?

vkuznet replied:
done

amaltaro reviewed (approved):
Valentin, this looks good to me. I'd suggest to add the docstring I mentioned in the code though.

In addition to that, please:
1) next time, please keep in mind that test/* changes should NOT go together with the source code changes, i.e., they must be provided in different commits;
2) please squash these commits in a single one (if test was separated, we should squash them in 2 commits instead).

Thanks

vkuznet commented:
@amaltaro now you have it squashed and rebased.

amaltaro commented:
test this please

todor-ivanov reviewed (commented):
looks good to me
